Add module, outline styles

yscik opened the pull request:
Apart from module & outline block styles and colors, this contains a lot of small style adjustment for various edge cases and combinations, generally aiming to support as much as of the theme's visuals as possible. 

### Changes proposed in this Pull Request

* Move border from Module to the parent Outline block — Looks like this was the design intention originally.
* Add module block styles (Default, Minimal, and an unstyled one for easy custom CSS)
* Add bordered/borderless outline block style
* Add outline & module colors
* Improve theme color support
* Adjust border colors in a few places to use text color
* Adjust lesson block style to design (whole block clickable), add a highlight for hover


### Testing instructions

* Create an outline with modules & lessons
* Customize appearance, check that everything looks the same on the frontend
* Customize colors using theme color palettes (the default colors in the Color Settings)
* On the frontend, Customize theme and check that changing theme colors applies to the outline colors too
* Check with a few themes

<!--
Helpful tips for screenshots:
https://en.support.wordpress.com/make-a-screenshot/
-->
### Screenshot / Video

<img width="607" alt="image" src="https://user-images.githubusercontent.com/176949/94318130-1a4b5080-ff88-11ea-9a8f-7f55978ca953.png">
<img width="607" alt="image" src="https://user-images.githubusercontent.com/176949/94318019-e07a4a00-ff87-11ea-954e-0178b1586013.png">

yscik commented on assets/blocks/course-outline/course-block/block.json:
```diff
@@ -12,6 +12,42 @@
+      "isDefault": true
     }
+  ],
+  "example": {
```
A block with these example attributes are also created for preview on the Block Style settings. (On addition to showing the example when inserting from the block library)

renatho replied:
Hum! Very interesting! TIL!

yscik commented on assets/blocks/course-outline/module-block/edit.js:
```diff
@@ -15,17 +20,21 @@ import { ModuleBlockSettings } from './settings';
-	setAttributes,
-} ) => {
+const EditModuleBlock = ( props ) => {
+	const {
+		className,
+		attributes: { title, description },
+		mainColor,
+		textColor,
+		setAttributes,
+		blockStyle,
+	} = props;
 	const {
 		outlineAttributes: { animationsEnabled },
-	} = useContext( OutlineAttributesContext );
-
+	} = useContext( OutlineAttributesContext ) || { outlineAttributes: {} };
```
Fixes an issue with the context not being available in the block preview.

yscik commented on assets/blocks/course-outline/module-block/edit.js:
```diff
@@ -122,14 +139,11 @@ const EditModuleBlock = ( {
+						</div>
 					</div>
 					<InnerBlocks
-						template={ [
-							[ 'sensei-lms/course-outline-lesson', {} ],
-						] }
```
Had to remove this for now as due to a Gutenberg bug, this template lesson block's inspector controls get added to the module's inspector panel each time a block style preview is rendered :|

yscik commented on assets/blocks/course-outline/style.scss:
```diff
@@ -54,68 +41,161 @@ $dark-gray: #1a1d20;
-	margin: 1.5em 0;
+	$block: '.wp-block-sensei-lms-course-outline-module';
+
+	&.is-style-default, &.is-style-minimal {
```
We only add style-related CSS when a style is set, to make it easier to turn off Sensei CSS in case someone wants to do their own thing.

yscik commented on includes/blocks/class-sensei-block-helpers.php:
```diff
@@ -73,10 +78,26 @@ public static function render_style_attributes( $class_names, $css ) {
+	public static function block_class_with_default_style( $attributes ) {
+		$class_name = $attributes['className'] ?? '';
+		if ( empty( $class_name ) && false === strpos( $class_name, 'is-style-' ) ) {
+			$class_name .= ' is-style-default';
```
The default style would not get added to the block class list, unless the user explicitly clicks it.

yscik commented on includes/blocks/class-sensei-course-outline-block.php:
```diff
@@ -248,29 +252,57 @@ private function render_module_block( $block, $course_id, $outline_attributes )
+				]
+			);
+
+			$style_header = '<div ' . Sensei_Block_Helpers::render_style_attributes( 'wp-block-sensei-lms-course-outline-module__name__minimal-border', $header_border_css ) . '></div>';
```
An additional element is used for the Minimal style module header border, so that it can use named theme colors via background-color, because classes for border color are usually not provided by themes.

renatho commented on assets/blocks/course-outline/module-block/index.js:
```diff
@@ -29,16 +28,57 @@ registerBlockType( 'sensei-lms/course-outline-module', {
+		attributes: {
+			title: 'Module',
+			description: 'About Module',
+			innerBlocks: [
```
I think it should be a sibling of the `attributes`, right? It didn't load the lesson in the preview, but in the Outline it worked.

yscik replied:
Ah, yes, thanks! And this also solves the issue with the innerblocks template! Fixing in 901943c

renatho reviewed (commented):
It's looking great!

Adding some comments from a first view.

yscik commented on assets/blocks/course-outline/style.scss:
```diff
@@ -1,7 +1,16 @@
+	border: 0 solid $dark-gray;
+
+	&.is-style-default {
+		border-width: 1px;
```
Yeah, good idea. 
I think we should add back the `Other lessons` header too for these lessons. I have a draft for that in #3658, will explore this there.

renatho commented on assets/shared/blocks/settings.js:
```diff
@@ -1,23 +1,89 @@
-				},
-			} ),
-		label,
+export function withColorSettings( colorSettings ) {
```
In some cases we're using arrow function and others a `function`. Did you use that here for some specific reason? If not, maybe we could update all for the same standard. And maybe we could also add this to our `eslint`. WDYT?

yscik replied:
Yeah we can do that. Updated in a few places in 888b321

renatho commented on assets/blocks/course-outline/course-block/edit.js:
```diff
@@ -74,26 +81,29 @@ const EditCourseOutlineBlock = ( {
+			<OutlineAttributesContext.Provider
+				value={ {
+					outlineAttributes: attributes,
+					outlineSetAttributes: setAttributes,
```
I checked that we're not using the `outlineSetAttributes` in any component, only the `outlineAttributes`. Did we already add it for a future feature?

Also, I think it's not part of this PR, but instead using the context directly, we could use the [block context](https://developer.wordpress.org/block-editor/developers/block-api/block-context/) for that. WDYT?

yscik replied:
That would be a good fit, but we likely want to use the setter too, and block context doesn't have that. Yeah the set was there for the future I think, using it for an upcoming shared settings PR.

renatho commented on assets/blocks/course-outline/module-block/edit.js:
```diff
@@ -122,9 +141,9 @@ const EditModuleBlock = ( {
 					</div>
 					<div className="wp-block-sensei-lms-course-outline-module__lessons-title">
-						<h3 className="wp-block-sensei-lms-course-outline__clean-heading">
+						<div className="wp-block-sensei-lms-course-outline__clean-heading">
```
I noticed that on the frontend it's without this `div`. And since it's no more an `h3`, I think we can remove this `div`.

yscik replied:
Changed the `Lessons` back to `h3` per your other comment, and normalized this bit in the editor. dc1bb75

yscik commented on assets/blocks/course-outline/style.scss:
```diff
@@ -54,68 +41,161 @@ $dark-gray: #1a1d20;
-	padding: 20px 25px;
+	margin: 1px 0;
+	position: relative;
+	font-size: 1rem;
```
I was trying to get away from `px` for things that would better be dictated by the theme, but good catch, this should be `em` at least. 

Stepping back, maybe it's better after all to not pick up the theme heading styles for `h2`, `h3`, since we have some extra styling around them. I'll try a solution where we have the theme defaults for the *Unstyled* variant, and add some stricter sizing to the other styles.

I'm a bit torn on whether we should even use `h3` for the lesson blocks, since they are not really headings in this context.

renatho replied:
It's likely impacting other parts using `rem` too.

renatho replied:
> I'm a bit torn on whether we should even use h3 for the lesson blocks, since they are not really headings in this context.

I agree with you! It's not title in this context. It doesn't need to be an `h3`.

renatho replied:
But in the case of the `Lessons` title, I think it should be a heading.

yscik replied:
Adjusted the headings mostly in dc1bb75
Also looks like the more generic overrides could be removed now 72f42bc

renatho replied:
I checked there are some more `rem` in spacings which can be impacted differently depending on the theme. I think we should replace all of them. Or did you keep them on purpose for some reason?

yscik replied:
I was thinking it's fine for the theme to provide, but probably not with root font sizes being set to weird values. Also not common in built-in component styles, so changed these to `px` (and put it in a variable where they should be the same for proper alignment)

renatho commented on assets/blocks/course-outline/style.scss:
```diff
@@ -54,68 +41,161 @@ $dark-gray: #1a1d20;
-	font-size: 16px;
 
-	&::before {
+	&:before {
```
Out of curiosity, did we changed `::` to `:` for some specific reason?

yscik replied:
TIL there is a pseudo-element syntax, thought it was a mistake :) Changed back in 31a41d8

renatho reviewed (commented):
Phew 😅. Now I finished a more complete review.

It's looking very nice! I just added a few more comments.

donnapep commented:
I tried to do some testing here but ran into an issue that isn't happening on `master`. I added 2 modules to the block, with 1 lesson each. I then went and published the 2 lessons. When I came back to the course, the modules had disappeared and only the lessons were left. This was on the Twenty Twenty theme.

For the bordered style, the module name ends up being tight to the margin. Also, if I add more modules, the style preview grows as well:

![Screen Shot 2020-10-02 at 8 36 08 AM](https://user-images.githubusercontent.com/1190420/94923736-556ae980-048a-11eb-8058-82d81065d972.jpg)

The module styles don't seem to match the original designs at all, where there were _Filled_ and _Minimal_ styles. Was this change discussed somewhere that I may have missed?

![Screen Shot 2020-10-02 at 8 38 38 AM](https://user-images.githubusercontent.com/1190420/94923967-b2669f80-048a-11eb-9ebd-b1e9e98468c9.jpg)

This PR also adds a _None_ style. Is that a typical design pattern?

Changing the background color of the first module didn't work for me; however, changing the background color of the second module did work. It wasn't clear that changing _Main color_ would update the background color though. Is it possible to rename it to _Background color_ for consistency?

yscik commented:
Looks like there was an issue with the last merge and some classname changes were left out. The 'Default' style was the filled one, it should work now. I also renamed it to 'Filled'. 

The 'None' style should make it easy to remove Sensei styling here, but yeah, this might not be a good place for it, removed it for now. 

Fixed some issues with the style previews, should be much clearer now.

I couldn't reproduce modules disappearing, this PR should not have anything affecting that.

> Main color / background color

This also sets the border-color for the minimal style header. Maybe *Header color*? I think Main color is used in a few places in the built-in blocks.

donnapep commented:
Those missing classes seem to have fixed the issues I was seeing with the module styles. Looks great!

One slight problem is that when the module style is _Minimal_, the collapsible module arrow disappears because it's white on white. Should probably be the same color as the border in that case.

For the module text color, when using the _Minimal_ style, it changes the color of the _In Progress_ status as well, but not when using _Filled_. Although I don't have a strong preference, I think I'd lean slightly towards not styling the module status, as that would leave the door open for potentially adding separate color settings for it in the future. Regardless, both module styles should behave the same for setting a text color.

> This also sets the border-color for the minimal style header. Maybe Header color? I think Main color is used in a few places in the built-in blocks.

Ah, I see. I'm good with either naming convention in that case.

I'm still seeing the issue with modules disappearing on this branch, but now I'm able to reliably reproduce:
- Create a course with a module and a lesson.
- Click _Save Draft_, which generates the following PHP notice:
```
PHP Notice:  Undefined index: id in /Users/donnapep/Local Sites/sensei/app/public/wp-content/plugins/sensei-lms/includes/blocks/class-sensei-course-outline-block.php on line 144
```
- Select the lesson block and click _Edit Lesson_.
- Publish the lesson.
- Refresh the course.
- Module is gone.

Similarly, if I open a course created on the `master` branch, then switch to this branch and refresh the course, I get this:

![Screen Shot 2020-10-05 at 7 34 40 AM](https://user-images.githubusercontent.com/1190420/95074878-3f089c00-06dd-11eb-8b3b-6e072ad4cfeb.jpg)

yscik commented:
Found the issue with the module save, the Block style previews messing things up again, re-enabling the legacy Course modules metaboxes 😞 

Fixed the arrow to also pick up the text color. For the module progress indicator, the filled status has a fixed black text on white background, which can be the inverse of the module header colors, so picking up the color settings wouldn't work there, it indeed needs it's own color settings to customize. But right now it works with any background. However, the Minimal style has a transparent background and border, so leaving it at a fixed color could cause issues, for example when the page has a dark background. So the safest bet is to pick up text color, for now at least. (Or keep it as a pill with white background)

donnapep commented:
Module disappearance issue looks to be fixed. Interestingly, after I select the _Course Modules_ and _Course Lessons_ metaboxes to display, next time I open the course they're gone again. So maybe it's not so bad.

After the build is fixed I can ✔️ this.
